test/libsolidity/SolidityNameAndTypeResolution.cpp

```diff
@@ -7313,6 +7313,59 @@ BOOST_AUTO_TEST_CASE(no_address_members_on_contract)
 	CHECK_ERROR(text, TypeError, "Member \"delegatecall\" not found or not visible after argument-dependent lookup in contract");
 }
 
+
+BOOST_AUTO_TEST_CASE(valid_hex_with_underscores)
+{
+	char const* text = R"(
+			contract test {
+				uint256 a = 0x1234_abcd;
+				uint256 b = 0x1234_5678_9abc;
+				uint256 c = 0x123_4567;
+				uint256 d = 0x1234_56 wei;
+				uint256 e = 0x12345678_9abc_abcd_abcd;
+                uint256 f = 0x1234_5678_9abc_abcdabcd;
+                uint256 x = 0x123456789abcdef;
+			}
+		)";
+		CHECK_SUCCESS(text);
+}
+
+BOOST_AUTO_TEST_CASE(hex_underscore_inconsistent_grouping)
+{
+	char const* text = R"(
+		contract C {
+			function f() {
+				uint256 a = 0xabcd_12345678_abcd;
+			}
+		}
+	)";
+	CHECK_ERROR(text, SyntaxError, "Invalid use of underscores in hex literal. Found inner part of length 8 (has to be 4 digits).");
```

Comment:
Uses `length` and `digits`. I think it should just use on term and `digits` seem to be fine. This applies to the other messages too.

Reply:
@chriseth this hasn't been addressed